{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define function to get spike sequences from each of the 50 states and write to file\n",
    "def getUSASpike():\n",
    "    #Import Biopython\n",
    "    from Bio import SeqIO\n",
    "    #Import OS\n",
    "    #import os\n",
    "    import re\n",
    "    #rbd = (r'RVQPTESIVRFPN')\n",
    "    #define rbd pattern\n",
    "    rbd = (r'[R][V][Q][A-Z]{0,223}[F]')\n",
    "    #create empty dict to hold record id and rbdseq for each state\n",
    "    rbdSeqs = {}\n",
    "    #define all states\n",
    "    states = [\"Alabama\",\"Alaska\",\"Arizona\",\"Arkansas\",\"California\",\"Colorado\",\"Connecticut\",\"Delaware\",\"District of Columbia\",\"Florida\",\"Georgia\",\"Hawaii\",\"Idaho\",\"Illinois\",\"Indiana\",\"Iowa\",\"Kansas\",\"Kentucky\",\"Louisiana\",\"Maine\",\"Montana\",\"Nebraska\",\"Nevada\",\"New Hampshire\",\"New Jersey\",\"New Mexico\",\"New York\",\"North Carolina\",\"North Dakota\",\"Ohio\",\"Oklahoma\",\"Oregon\",\"Maryland\",\"Massachusetts\",\"Michigan\",\"Minnesota\",\"Mississippi\",\"Missouri\",\"Pennsylvania\",\"Rhode Island\",\"South Carolina\",\"South Dakota\",\"Tennessee\",\"Texas\",\"Utah\",\"Vermont\",\"Virginia\",\"Washington\",\"West Virginia\",\"Wisconsin\",\"Wyoming\"]\n",
    "    #iter through states\n",
    "    for i in states:\n",
    "        #define file path\n",
    "        file = f\"/Users/Kwaz9/Desktop/GradSchool/ProblemSolvingBioinformatics/FinalProject/Spike50States/50States_Spike.fasta\"\n",
    "        #Create file with open and use append functionality\n",
    "        with open(file,\"a+\") as myfile:\n",
    "            #open allsurfaceglucoproteinseqs.fasta as handle\n",
    "            with open('./AllSurfaceGlycoproteinSeqs.fasta') as handle:\n",
    "                #parse fasta file and store in variable\n",
    "                records = list(SeqIO.parse(handle,'fasta'))\n",
    "                #for record in list(SeqIO.parse(handle, \"fasta\")):\n",
    "                    #find each state in records\n",
    "                for record in records[:101]:\n",
    "                    if i in record.description:\n",
    "                        match = re.search(rbd, str(record.seq))\n",
    "                        if match:\n",
    "                            myfile.write('>'+record.id+ ' '+ record.description+'\\n')\n",
    "                            myfile.write(match[0]+'\\n')\n",
    "                                               \n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define function to get spike sequences from all USA regions and write to file\n",
    "def getRegionSpike():\n",
    "    #Import Biopython\n",
    "    from Bio import SeqIO\n",
    "    #Import OS\n",
    "    import os\n",
    "    #improt regex\n",
    "    import re\n",
    "    #define rbd regex\n",
    "    rbd = (r'[R][V][Q][A-Z]{0,223}[F]')\n",
    "    #define all reions\n",
    "    regions = {'northEast': ['Connecticut', 'Maine', 'Massachusetts', 'New Hampshire', 'Rhode Island', 'Vermont','New Jersey', 'New York', 'Pennsylvania'],\n",
    "                 'southEast': ['Alabama', 'Kentucky', 'Mississippi', 'Tennessee'],\n",
    "              'midWest': ['Illinois', 'Indiana', 'Michigan', 'Ohio', 'Wisconsin', 'Iowa', 'Kansas', 'Minnesota', 'Missouri', 'Nebraska', 'North Dakota', 'South Dakota'],\n",
    "              'mountain': ['Colorado', 'Wyoming', 'Utah', 'New Mexico', 'Nevada', 'Idaho', 'Arizona', 'Montana'],\n",
    "              'southWest': ['Arkansas', 'Louisiana', 'Oklahoma', 'Texas'],\n",
    "               'pacific': ['Alaska', 'Hawaii', 'California', 'Oregon', 'Washington'],\n",
    "              }\n",
    "\n",
    "    #iter through states by region\n",
    "    for k, v in regions.items():\n",
    "        #define file path\n",
    "        file = f\"/Users/Kwaz9/Desktop/GradSchool/ProblemSolvingBioinformatics/FinalProject/SpikeByRegion/{k}_Spike.fasta\"\n",
    "        #create file with open\n",
    "        with open(file,\"a+\") as myfile:\n",
    "           #open seq.fasta as handle\n",
    "            with open('./AllSurfaceGlycoproteinSeqs.fasta') as handle:\n",
    "                #parse fasta file\n",
    "                records = list(SeqIO.parse(handle,\"fasta\"))\n",
    "                #for record in list(SeqIO.parse(handle, \"fasta\")):\n",
    "                for record in records[:101]:\n",
    "                    #iter through values of each key (region)\n",
    "                    for i in v:\n",
    "                       #find state in records\n",
    "                        if i in record.description:\n",
    "                            #if seq not in file write to file\n",
    "                            match = re.search(rbd, str(record.seq))\n",
    "                            if match:\n",
    "                                myfile.write('>'+record.id+ ' '+ record.description+'\\n')\n",
    "                                myfile.write(match[0]+'\\n')\n",
    "\n",
    "                \n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define function to get spike sequences from all USA regions and write to file\n",
    "def getRegionSpike():\n",
    "    #Import Biopython\n",
    "    from Bio import SeqIO\n",
    "    #Import OS\n",
    "    import os\n",
    "    #improt regex\n",
    "    import re\n",
    "    #define rbd regex\n",
    "    rbd = (r'[R][V][Q][A-Z]{0,223}[F]')\n",
    "    #define all reions\n",
    "    regions = {'northEast': ['Connecticut', 'Maine', 'Massachusetts', 'New Hampshire', 'Rhode Island', 'Vermont','New Jersey', 'New York', 'Pennsylvania'],\n",
    "                 'southEast': ['Alabama', 'Kentucky', 'Mississippi', 'Tennessee'],\n",
    "              'midWest': ['Illinois', 'Indiana', 'Michigan', 'Ohio', 'Wisconsin', 'Iowa', 'Kansas', 'Minnesota', 'Missouri', 'Nebraska', 'North Dakota', 'South Dakota'],\n",
    "              'mountain': ['Colorado', 'Wyoming', 'Utah', 'New Mexico', 'Nevada', 'Idaho', 'Arizona', 'Montana'],\n",
    "              'southWest': ['Arkansas', 'Louisiana', 'Oklahoma', 'Texas'],\n",
    "               'pacific': ['Alaska', 'Hawaii', 'California', 'Oregon', 'Washington'],\n",
    "              }\n",
    "#     regions = {'northEast':['Connecticut', 'Maine', 'Massachusetts', 'New Hampshire', 'Rhode Island', 'Vermont','New Jersey', 'New York', 'Pennsylvania'],\n",
    "#                 'midWest':['Illinois', 'Indiana', 'Michigan', 'Ohio', 'Wisconsin', 'Iowa', 'Kansas', 'Minnesota', 'Missouri', 'Nebraska', 'North Dakota', 'South Dakota'],\n",
    "#                 'south': ['Delaware', 'Florida', 'Georgia', 'Maryland','North Carolina', 'South Carolina', 'Virginia', \"District of Columbia\", 'West Virginia', 'Alabama', 'Kentucky', 'Mississippi', 'Tennessee','Arkansas', 'Louisiana', 'Oklahoma', 'Texas'],\n",
    "#                 'west': ['Arizona', 'Colorado', 'Idaho', 'Montana', 'Nevada', 'New Mexico', 'Utah', 'Wyoming', 'Alaska', 'California', 'Hawaii', 'Oregon', 'Washington']\n",
    "#                   }\n",
    "    \n",
    "#     with open('./AllSurfaceGlycoproteinSeqs.fasta') as handle:\n",
    "#         #parse fasta file\n",
    "#         records = list(SeqIO.parse(handle,\"fasta\"))\n",
    "    #iter through states by region\n",
    "    for k, v in regions.items():\n",
    "        #iter through values \n",
    "        for i in v:\n",
    "            #define file path\n",
    "            file = f\"/Users/Kwaz9/Desktop/GradSchool/ProblemSolvingBioinformatics/FinalProject/SpikeByRegion/{k}_Spike.fasta\"\n",
    "            state_file = f\"/Users/Kwaz9/Desktop/GradSchool/ProblemSolvingBioinformatics/FinalProject/Spike50States/{i}_Spike.fasta\"\n",
    "            #create file with open\n",
    "            with open(file,\"a+\") as myfile:\n",
    "                #open seq.fasta as handle for each individual state\n",
    "                with open(state_file) as handle:\n",
    "                    records = list(SeqIO.parse(handle,\"fasta\"))\n",
    "                #for record in list(SeqIO.parse(handle, \"fasta\")):\n",
    "                    for record in records[:101]:\n",
    "                        #iter through values of each key (region)\n",
    "                    #for i in v:\n",
    "                        #count = 0\n",
    "                        #while count < 101:\n",
    "                           #find state in records\n",
    "                        if i in record.description:\n",
    "                                #count+=1\n",
    "                            #if seq not in file write to file\n",
    "                            match = re.search(rbd, str(record.seq))\n",
    "                            if match:\n",
    "                                myfile.write('>'+record.id+ ' '+ record.description+'\\n')\n",
    "                                myfile.write(match[0]+'\\n')\n",
    "#                             if record.seq not in myfile:\n",
    "#                                 #write to file\n",
    "#                                 SeqIO.write(record, myfile, \"fasta\")\n",
    "                \n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define function to get spike sequences from states with no mask mandate and write to file\n",
    "def getNoMaskSpike():\n",
    "    #Import Biopython\n",
    "    from Bio import SeqIO\n",
    "    #Import OS\n",
    "    import os\n",
    "    #import regex\n",
    "    import re\n",
    "    #define rbd regex\n",
    "    rbd = (r'[R][V][Q][A-Z]{0,223}[F]')\n",
    "    #define all no mask mandate states\n",
    "    noMask = ['Arizona', 'Florida', 'Georgia', 'Idaho', 'Missouri', 'Nebraska',\n",
    "'Oklahoma', 'South Carolina', 'South Dakota', 'Tennessee']\n",
    "    #iter through states\n",
    "    for i in noMask:\n",
    "        #define file path\n",
    "        file = f\"/Users/Kwaz9/Desktop/GradSchool/ProblemSolvingBioinformatics/FinalProject/SpikeNoMask/NoMask_Spike.fasta\"\n",
    "        state_file = f\"/Users/Kwaz9/Desktop/GradSchool/ProblemSolvingBioinformatics/FinalProject/Spike50States/{i}_Spike.fasta\"\n",
    "        #create file with open and use append functionality\n",
    "        with open(file,\"a+\") as myfile:\n",
    "         #open seqs.fasta as handle\n",
    "            with open(state_file) as handle:\n",
    "                #parse fasta file\n",
    "                records = list(SeqIO.parse(handle,\"fasta\"))\n",
    "                for record in records[:101]:\n",
    "                #for record in list(SeqIO.parse(handle, \"fasta\")):\n",
    "                    #find each state in records\n",
    "                    if i in record.description:\n",
    "                        #search record sequence for rbd sequence\n",
    "                        match = re.search(rbd, str(record.seq))\n",
    "                        #if match\n",
    "                        if match:\n",
    "                            #write to file\n",
    "                            myfile.write('>'+record.id+ ' '+ record.description+'\\n')\n",
    "                            myfile.write(match[0]+'\\n')\n",
    "#                         #write to file\n",
    "#                         SeqIO.write(record, myfile, \"fasta\")\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define function to get spike protein of states with limites size of outdoor gatherings\n",
    "\n",
    "def getLimitedOutdoorSpike():\n",
    "    #Import Biopython\n",
    "    from Bio import SeqIO\n",
    "    #Import OS\n",
    "    import os\n",
    "    #import regex\n",
    "    import re\n",
    "    #define rbd regex\n",
    "    rbd = (r'[R][V][Q][A-Z]{0,223}[F]')\n",
    "    #define all limited outdoor states\n",
    "    limitedOutdoor = ['California', 'Colorado','Connecticut', 'Delaware', 'Hawaii',\n",
    "'Maine', 'Massachusetts', 'Michigan', 'New Jersey', 'New York', 'North Carolina', 'Pennsylvania',\n",
    "'Vermont']\n",
    "    #iter through states\n",
    "    for i in limitedOutdoor:\n",
    "        #define file path\n",
    "        file = f\"/Users/Kwaz9/Desktop/GradSchool/ProblemSolvingBioinformatics/FinalProject/SpikeLimtedOutdoor/LimtedOutdoor_Spike.fasta\"\n",
    "        state_file = f\"/Users/Kwaz9/Desktop/GradSchool/ProblemSolvingBioinformatics/FinalProject/Spike50States/{i}_Spike.fasta\"\n",
    "        #create file with open and use append functionality\n",
    "        with open(file,\"a+\") as myfile:\n",
    "        #open seqs.fasta as handle\n",
    "            with open(state_file) as handle:\n",
    "                #parse fasta file\n",
    "                records = list(SeqIO.parse(handle,\"fasta\"))\n",
    "                #for record in list(SeqIO.parse(handle, \"fasta\")):\n",
    "                for record in records[:101]:   \n",
    "                    #find each state in records\n",
    "                    if i in record.description:\n",
    "                        #search record sequence for rbd sequence\n",
    "                        match = re.search(rbd, str(record.seq))\n",
    "                        #if match\n",
    "                        if match:\n",
    "                            #write to file\n",
    "                            myfile.write('>'+record.id+ ' '+ record.description+'\\n')\n",
    "                            myfile.write(match[0]+'\\n')                    \n",
    "    return\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create function to get spike protein of highest partial vaccination\n",
    "\n",
    "def getHighestVaxSpike():\n",
    "    #Import Biopython\n",
    "    from Bio import SeqIO\n",
    "    #Import OS\n",
    "    import os\n",
    "    #import re\n",
    "    import re\n",
    "    #define rbd regex\n",
    "    rbd = (r'[R][V][Q][A-Z]{0,223}[F]')\n",
    "    #define all states\n",
    "    highestPartial = ['Vermont', 'Rhode Island', 'Maine', 'Connecticut', 'Massachusetts', 'New York', 'New Jersey', 'Maryland', 'Virginia', 'Washington', 'District of Columbia', 'New Hampshire', 'Oregon', 'New Mexico',\n",
    "'California', 'Colorado', 'Minnesota']\n",
    "    #iter through states\n",
    "    for i in highestPartial:\n",
    "        #define file path\n",
    "        file = f\"/Users/Kwaz9/Desktop/GradSchool/ProblemSolvingBioinformatics/FinalProject/SpikeHighestVax/HighestVax_Spike.fasta\"\n",
    "        state_file = f\"/Users/Kwaz9/Desktop/GradSchool/ProblemSolvingBioinformatics/FinalProject/Spike50States/{i}_Spike.fasta\"\n",
    "        #create file with open and use append functionality\n",
    "        with open(file,\"a+\") as myfile:\n",
    "       #open seqs.fasta as handle\n",
    "            with open(state_file) as handle:\n",
    "                #parse fasta file\n",
    "                records = list(SeqIO.parse(handle,\"fasta\"))\n",
    "                for record in records[:101]:\n",
    "                #for record in list(SeqIO.parse(handle, \"fasta\")):\n",
    "                    #find each state in records\n",
    "                    if i in record.description:\n",
    "                        #search record sequence for rbd sequence\n",
    "                        match = re.search(rbd, str(record.seq))\n",
    "                        #if match\n",
    "                        if match:\n",
    "                            #write to file\n",
    "                            myfile.write('>'+record.id+ ' '+ record.description+'\\n')\n",
    "                            myfile.write(match[0]+'\\n') \n",
    "    \n",
    "    return\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create function to get spike protein of lowest partial vaccination\n",
    "def getLowestPartialSpike():\n",
    "    #Import Biopython\n",
    "    from Bio import SeqIO\n",
    "    #Import OS\n",
    "    import os\n",
    "    #import regex\n",
    "    import re\n",
    "    #define rbd regex\n",
    "    rbd = (r'[R][V][Q][A-Z]{0,223}[F]')\n",
    "    #define all states\n",
    "    lowestPartial = ['Idaho', 'Wyoming', 'Mississippi', 'West Virginia', 'Louisiana',\n",
    "'Alabama', 'Indiana', 'Tennessee', 'North Dakota', 'Ohio']\n",
    "    #iter through states\n",
    "    for i in lowestPartial:\n",
    "        #define file path\n",
    "        file = f\"/Users/Kwaz9/Desktop/GradSchool/ProblemSolvingBioinformatics/FinalProject/SpikeLowestPartial/LowestPartial_Spike.fasta\"\n",
    "        state_file = f\"/Users/Kwaz9/Desktop/GradSchool/ProblemSolvingBioinformatics/FinalProject/Spike50States/{i}_Spike.fasta\"\n",
    "        #create file with open and use append functionality\n",
    "        with open(file,\"a+\") as myfile:\n",
    "            #open fasta as handle\n",
    "            with open(state_file) as handle:\n",
    "                #parse fasta file\n",
    "                records = list(SeqIO.parse(handle,\"fasta\"))\n",
    "                for record in records[:101]:\n",
    "                    #print(record)\n",
    "                #for record in list(SeqIO.parse(handle, \"fasta\")):\n",
    "                    #find each state in records\n",
    "                    if i in record.description:\n",
    "                        #search record sequence for rbd sequence\n",
    "                        match = re.search(rbd, str(record.seq))\n",
    "                        #if match\n",
    "                        if match:\n",
    "                            #write to file\n",
    "                            myfile.write('>'+record.id+ ' '+ record.description+'\\n')\n",
    "                            myfile.write(match[0]+'\\n') \n",
    "                        #SeqIO.write(record, myfile, \"fasta\")\n",
    "    return\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Call getUSAspike\n",
    "getUSASpike()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#call getRegionSpike\n",
    "getRegionSpike()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#call getLowestPartialSpike\n",
    "getLowestPartialSpike()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#call getLimitedOutdoorSpike\n",
    "getLimitedOutdoorSpike()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#call getHighestVaxSpike\n",
    "getHighestVaxSpike()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#call getNoMaskSpike\n",
    "getNoMaskSpike()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "#write function to do multiple sequence alignment\n",
    "def alignRegions():\n",
    "    #import os\n",
    "    import os\n",
    "    #define list of fasta files for each region\n",
    "    regions = {\n",
    "    'northEast':'C:\\\\Users\\\\Kwaz9\\\\Desktop\\\\GradSchool\\\\ProblemSolvingBioinformatics\\\\FinalProject\\\\SpikeByRegion\\\\northEast_Spike.fata',\n",
    "    'southEast':'C:\\\\Users\\\\Kwaz9\\\\Desktop\\\\GradSchool\\\\ProblemSolvingBioinformatics\\\\FinalProject\\\\SpikeByRegion\\\\southEast_Spike.fasta',\n",
    "    'midWest':'C:\\\\Users\\\\Kwaz9\\\\Desktop\\\\GradSchool\\\\ProblemSolvingBioinformatics\\\\FinalProject\\\\SpikeByRegion\\\\midWest_Spike.fasta',\n",
    "    'mountain':'C:\\\\Users\\\\Kwaz9\\\\Desktop\\\\GradSchool\\\\ProblemSolvingBioinformatics\\\\FinalProject\\\\SpikeByRegion\\\\mountain_Spike.fasta',\n",
    "    'southWest':'C:\\\\Users\\\\Kwaz9\\\\Desktop\\\\GradSchool\\\\ProblemSolvingBioinformatics\\\\FinalProject\\\\SpikeByRegion\\\\southWest_Spike.fasta',\n",
    "    'pacific':'C:\\\\Users\\\\Kwaz9\\\\Desktop\\\\GradSchool\\\\ProblemSolvingBioinformatics\\\\FinalProject\\\\SpikeByRegion\\\\pacific_Spike.fasta'\n",
    "    }\n",
    "    #iter through regions dictionary\n",
    "    for k, v in regions.items():\n",
    "        #define command\n",
    "        cmd = f\".\\\\clustal-omega-1.2.2-win64\\\\clustalo.exe --in {v} --out .\\\\Alignment\\\\{k}_alignment.fasta\"\n",
    "        #run command with os.system\n",
    "        os.system(cmd)\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def alignNoMask():\n",
    "    import os\n",
    "    #define no mask path\n",
    "    noMask = f\"\\\\Users\\\\Kwaz9\\\\Desktop\\\\GradSchool\\\\ProblemSolvingBioinformatics\\\\FinalProject\\\\SpikeNoMask\\\\NoMask_Spike.fasta\"\n",
    "    #define command\n",
    "    cmd = f\".\\\\clustal-omega-1.2.2-win64\\\\clustalo.exe --in {noMask} --out .\\\\Alignment\\\\NoMask_alignment.fasta\"\n",
    "    #call command\n",
    "    os.system(cmd)\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def alignLowestPartial():\n",
    "    import os\n",
    "    #define lowest partial seq path\n",
    "    lowestPartial = f\"\\\\Users\\\\Kwaz9\\\\Desktop\\\\GradSchool\\\\ProblemSolvingBioinformatics\\\\FinalProject\\\\SpikeLowestPartial\\\\LowestPartial_Spike.fasta\"\n",
    "    #define command\n",
    "    cmd = f\".\\\\clustal-omega-1.2.2-win64\\\\clustalo.exe --in {lowestPartial} --out .\\\\Alignment\\\\lowestPartial_alignment.fasta\"\n",
    "    #call command\n",
    "    os.system(cmd)\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def alignHighestVax():\n",
    "    import os\n",
    "    #define highest partial seq path\n",
    "    highestVax = f\"\\\\Users\\\\Kwaz9\\\\Desktop\\\\GradSchool\\\\ProblemSolvingBioinformatics\\\\FinalProject\\\\SpikeHighestVax\\\\HighestVax_Spike.fasta\"\n",
    "    #define command\n",
    "    cmd = f\".\\\\clustal-omega-1.2.2-win64\\\\clustalo.exe --in {highestVax} --out .\\\\Alignment\\\\HighestVax_alignment.fasta\"\n",
    "    #call command\n",
    "    os.system(cmd)\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def alignLimitedOutdoor():\n",
    "    import os\n",
    "    #define limited outdoor seq path\n",
    "    limitedOutdoor = f\"\\\\Users\\\\Kwaz9\\\\Desktop\\\\GradSchool\\\\ProblemSolvingBioinformatics\\\\FinalProject\\\\SpikeLimtedOutdoor\\\\LimtedOutdoor_Spike.fasta\"\n",
    "    #define command\n",
    "    cmd = f\".\\\\clustal-omega-1.2.2-win64\\\\clustalo.exe --in {limitedOutdoor} --out .\\\\Alignment\\\\limitedOutdoor_alignment.fasta\"\n",
    "    #call command\n",
    "    os.system(cmd)\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "#call alignRegions\n",
    "alignRegions()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "#call alignLowestParital\n",
    "alignLowestPartial()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#call alignNoMask\n",
    "alignNoMask()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "#call LimitedOutdoor align\n",
    "alignLimitedOutdoor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "#call alighnHighestVax\n",
    "alignHighestVax()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define funciton to convert fasta to Nexus file\n",
    "def convertFastaToNexus():\n",
    "    #import phylo and SeqIO\n",
    "    from Bio import Phylo, SeqIO\n",
    "    #define list of regions\n",
    "    filenames = ['northEast','southEast','midWest','mountain','southWest','pacific', \"NoMask\", \"HighestVax\", \"limitedOutdoor\", \"lowestPartial\"]\n",
    "    #iter through filenames\n",
    "    for i in filenames:\n",
    "        #define file path \n",
    "        file = f\"./Alignment/{i}_alignment.fasta\"\n",
    "        out = f\"./Alignment/{i}_alignment.nexus\"\n",
    "        #convert MSA fasta to nexus format\n",
    "        SeqIO.convert(file, \"fasta\", out,\"nexus\",\"protein\")\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generatePhylo():\n",
    "    #import phylo and AlignIO\n",
    "    from Bio import Phylo, AlignIO\n",
    "    #import Distance Calculator and DistanceTreeConstructor\n",
    "    from Bio.Phylo.TreeConstruction import DistanceCalculator, DistanceTreeConstructor\n",
    "    #define filenames list\n",
    "    filenames = ['northEast','southEast','midWest','mountain','southWest','pacific', \"NoMask\", \"HighestVax\", \"limitedOutdoor\", \"lowestPartial\"]\n",
    "    #iter through filenames\n",
    "    for i in filenames[7:8]:\n",
    "        #define file path\n",
    "        file = f\"./Alignment/{i}_alignment.nexus\"\n",
    "        #create MSA alignment object\n",
    "        aln = AlignIO.read(file, 'nexus')\n",
    "        #define calculator \n",
    "        calc = DistanceCalculator('identity')\n",
    "        #calculate distnace measurement\n",
    "        dm = calc.get_distance(aln)\n",
    "        #define constructor\n",
    "        constructor = DistanceTreeConstructor(calc, 'nj')\n",
    "        trees = constructor.build_tree(aln)\n",
    "        #generate phylogenetic tree\n",
    "        #create path to write tree files\n",
    "        out = f\"./Trees/{i}_phylo.xml\"\n",
    "        #write trees to new xml file\n",
    "        Phylo.write(trees, out, 'phyloxml')\n",
    "\n",
    "    return\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#write function to draw trees\n",
    "def drawTrees():\n",
    "    #import phylo\n",
    "    from Bio import Phylo\n",
    "    import matplotlib\n",
    "    import matplotlib.pyplot as plt\n",
    "    #from Phylo import Phylo.PhyloXMLIO\n",
    "    #define file names\n",
    "    filenames = ['northEast','southEast','midWest','mountain','southWest','pacific', \"NoMask\", \"HighestVax\", \"limitedOutdoor\", \"lowestPartial\"]\n",
    "    #iter through filenames\n",
    "    for i in filenames:\n",
    "        #define file path\n",
    "        file = f\"./Trees/{i}_phylo.xml\"\n",
    "        #read each tree\n",
    "        #trees = Phylo.PhyloXMLIO.parse(file)\n",
    "        trees = Phylo.read(file,'phyloxml')\n",
    "        #draw trees\n",
    "        #trees.ladderize()\n",
    "        #fig = plt.figure(figsize=(20,20), dpi=100)\n",
    "        #axes = fig.add_subplot(1,1,1)  \n",
    "        fig = plt.figure(figsize=(50,500), dpi=100) \n",
    "        matplotlib.rc('font', size=12)              \n",
    "        matplotlib.rc('xtick', labelsize=10)       \n",
    "        matplotlib.rc('ytick', labelsize=10)       \n",
    "        #turtle_tree.ladderize()\t\t    \n",
    "        axes = fig.add_subplot(1, 1, 1)\n",
    "        Phylo.draw(trees, axes=axes, branch_labels=lambda x: x.name)\n",
    "        #print(trees)\n",
    "        fig.savefig(f\"./Trees/{i}\")\n",
    "        \n",
    "        #plt.show()\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "#call regions convert\n",
    "convertFastaToNexus()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#call generatePhylo()\n",
    "generatePhylo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 5000x50000 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 5000x50000 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Draw trees\n",
    "drawTrees()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
